【OJT最終課題 ECサイトの作成】クラス図/シーケンス図 UC01[商品を一覧表示する]部分

wnakayama opened the pull request:
**OJT課題 概要、機能仕様書**
以下のmarkdownファイルにまとめました。
https://github.com/W-Nakayama-24/ojt_ecsite/blob/master/README.md

**ご確認いただきたい部分**
- シーケンスは一般的な非同期データ通信の形式になっているか
- クラス設計はオブジェクト指向に近づけているか (他のユースケースに使い回しづらい箇所、値の変更に弱い箇所は無いか)

**成果物**
https://github.com/W-Nakayama-24/ojt_ecsite/pull/3/files
- クラス図 UML/ClassDiag.puml


![](https://raw.githubusercontent.com/W-Nakayama-24/ojt_ecsite/bc77071bca44d02e2cf9251cc1d2d4b1c24b24ff/UML/EC%E3%82%B5%E3%82%A4%E3%83%88_%E3%82%AF%E3%83%A9%E3%82%B9%E5%9B%B3.png)


- シーケンス図 UML/Sequence01.puml


![](https://raw.githubusercontent.com/W-Nakayama-24/ojt_ecsite/bc77071bca44d02e2cf9251cc1d2d4b1c24b24ff/UML/EC%E3%82%B5%E3%82%A4%E3%83%88_UC01%E5%95%86%E5%93%81%E3%82%92%E4%B8%80%E8%A6%A7%E8%A1%A8%E7%A4%BA%E3%81%99%E3%82%8B.png)

wnakayama committed:
create classDiag, sequence01

wnakayama committed:
fix sequence01

iwanam commented:
- 商品が税率を持つ必要はあるかな？
- Teamsでもちょっと言ったけど、税込み価格だけを合計すると誤差が出るよ
- アイコン的な問題かもしれないけど、商品データのjsonは1ファイルで作ってね
- 同期非同期は使い分けられている？　Productに渡しているのはどちら？
- 完全な商品一覧とは？気持ちわかるけど

wnakayama committed:
商品データ.jsonのアイコンを変更, noteを一部修正

wnakayama committed:
Productクラスから消費税計算を別クラスに分離

wnakayama committed:
余分なコメントを削除

tmatsumura commented on UML/Sequence01.puml:
```diff
@@ -0,0 +1,63 @@
+元のJSONデータに「税込み価格」は無いので、
+Productクラス内のメソッドで税込み価格を新しく算出する
+end note
+ViewAllProductServlet <- Product : Productオブジェクトを返却(商品ID, 商品名, 商品カテゴリ, 税抜き価格, **税込み価格**)
```
レスポンスの場合は点線(`<--`)ではないですかね？
- http://www.itsenka.com/contents/development/uml/sequence.html

wnakayama replied:
忘れていました。。。
ご指摘ありがとうございます。直しておきます

tmatsumura replied:
レスポンス以外は大丈夫ですか？同期・非同期でも線が変わると思いますが、問題なければOKです

wnakayama replied:
ありがとうございます。
非同期通信部分のライフラインと矢印の表記を修正しました。 https://github.com/W-Nakayama-24/ojt_ecsite/pull/3/commits/91bfb5d161a5373029428e72073486380efbd3f2

![image](https://user-images.githubusercontent.com/54921025/72965795-94777400-3e00-11ea-9aa0-a4b55780299f.png)

tmatsumura reviewed (changes requested):
- 商品画像のデータはどこにあって、どういうふうにブラウザに渡しますか？
- 表示する順番は価格が安い順と書いてありますが、取得した複数のProductオブジェクトを並べ替える必要はないですか？

iwanam commented:
ふと思ったけど、jsonじゃなくてcsvは一つにしてね、だったわ・・・

W-Nakayama-24 commented:
> Teamsでもちょっと言ったけど、税込み価格だけを合計すると誤差が出るよ

口頭でも相談させていただきましたが、①の手法で開発しようと思います。
「税込み価格」のデータ型や「税込み価格カリキュレータ」の処理に変更が必要なため後ほどコミットします。

![image](https://user-images.githubusercontent.com/54921025/72860867-88f65100-3d0b-11ea-9d85-26e52d67786c.png)

wnakayama committed:
処理の追加、メッセージの修正

*商品データを価格が安い順に並び替える処理を追加
*レスポンスの矢印を点線に修正
*税込み価格の算出方法, 小数点以下切り捨てのタイミングを設定 データ型の修正

W-Nakayama-24 commented:
現在のクラス図/シーケンス図を改めてこちらに掲載します。

![](https://raw.githubusercontent.com/W-Nakayama-24/ojt_ecsite/e743a2506e24a8ae11dc833dc0f5b0828ff9e59f/UML/EC%E3%82%B5%E3%82%A4%E3%83%88_%E3%82%AF%E3%83%A9%E3%82%B9%E5%9B%B3.png)

![](https://raw.githubusercontent.com/W-Nakayama-24/ojt_ecsite/e743a2506e24a8ae11dc833dc0f5b0828ff9e59f/UML/EC%E3%82%B5%E3%82%A4%E3%83%88_UC01%E5%95%86%E5%93%81%E3%82%92%E4%B8%80%E8%A6%A7%E8%A1%A8%E7%A4%BA%E3%81%99%E3%82%8B.png)

W-Nakayama-24 commented:
>商品画像のデータはどこにあって、どういうふうにブラウザに渡しますか？

以下のようなディレクトリ構造をイメージしており、画像データはimagesディレクトリの配下に置く予定です。

![image](https://user-images.githubusercontent.com/54921025/72868936-e77cf880-3d26-11ea-9d08-2cd491a7d259.png)

画像データをブラウザに表示させるまでの動きは、後ほどシーケンス図に明記します。

iwanam commented:
- `jsonファイル` ではないよ
- シーケンス図の矢印はメソッドを表すので、ObjectMapper.readValueはおかしい
- 自作ライブラリにJSON形式で返すメソッドはないのでは？
- Productがデータを表すクラスだとしたら、そこから処理を呼び出す？
- サイトのディレクトリ構造じゃなくて、それはEclipseの作業フォルダの構造では？
  - Webサーバに置いたものは大体見られるので、.javaを置くとマズイよね
  - 難しかったら飛ばしてもいいけど、Webアプリの配置はある程度決まっているので

wnakayama committed:
クラス設計,シーケンス図のメッセージ,note部分の記述を修正

W-Nakayama-24 commented:
>表示する順番は価格が安い順と書いてありますが、取得した複数のProductオブジェクトを並べ替える必要はないですか？

>jsonファイル ではないよ

>シーケンス図の矢印はメソッドを表すので、ObjectMapper.readValueはおかしい

>自作ライブラリにJSON形式で返すメソッドはないのでは？

>Productがデータを表すクラスだとしたら、そこから処理を呼び出す？

いただいた5点の指摘に対応しました。


以下の2点は検討中です。

>サイトのディレクトリ構造じゃなくて、それはEclipseの作業フォルダの構造では？
確かにEclipseの作業フォルダを参考にしていました。
適切なWebサイトのディレクトリ構造を一度調べておきます。

>商品画像のデータはどこにあって、どういうふうにブラウザに渡しますか？

jQueryのセレクタを使って`<p id = 'product1'></p>` や `<div class = 'image'></div>`といった箇所を指定し、画像データのURLを指定したimgタグを追加するところまではイメージしていますが、実際にjavaScriptのコーディングに移ってみないと明確にお答えしづらい状態です。

W-Nakayama-24 commented:
**クラス図**
![](https://raw.githubusercontent.com/W-Nakayama-24/ojt_ecsite/5db5fb9332ae0b82f68fa25fa158c957eeee44f4/UML/EC%E3%82%B5%E3%82%A4%E3%83%88_%E3%82%AF%E3%83%A9%E3%82%B9%E5%9B%B3.png)

**シーケンス図**
![](https://raw.githubusercontent.com/W-Nakayama-24/ojt_ecsite/5db5fb9332ae0b82f68fa25fa158c957eeee44f4/UML/EC%E3%82%B5%E3%82%A4%E3%83%88_UC01%E5%95%86%E5%93%81%E3%82%92%E4%B8%80%E8%A6%A7%E8%A1%A8%E7%A4%BA%E3%81%99%E3%82%8B.png)

W-Nakayama-24 commented:
質問が2点あります


![image](https://user-images.githubusercontent.com/54921025/72949991-514fdd80-3dcd-11ea-960a-54062e4fa1ad.png)

- 画像データが該当する商品データの上部に正しく表示されるためにはどうするのが適切か分からないので、ヒントを頂けると助かります。。

  初学者向けの記事や大手ECサイトのコードを見てイメージしているのは以下のいずれかの方法をとることです
  
  - 商品データの商品IDと、画像データのhtmlタグのIDを同じにして管理する

  - `<div class = "product01"></div>`を使って 
商品情報[ { 画像データ } { 商品データ (商品名,税抜き価格, 税込み価格) } ]
という構造にまとめる

- 画像データをブラウザに表示させる作業は、HTML, javaScriptのプログラミングだけで完結するという認識で合っているでしょうか?

tmatsumura commented:
> 画像データが該当する商品データの上部に正しく表示されるためにはどうするのが適切か分からないので、ヒントを頂けると助かります。。

画像のURLと商品の対応付けができていればどの方法でも問題ないと思います。
例えば、一番左に表示する商品の情報が分かっていれば、それに対応するURLをセットしたimgタグを追加すればできそうな気はします

> 画像データをブラウザに表示させる作業は、HTML, javaScriptのプログラミングだけで完結するという認識で合っているでしょうか?

はい、javascriptでhtmlタグを追加していくのであればjavascriptでできるんじゃないかなと思います

W-Nakayama-24 commented:
ディレクトリ構造を考え直しました。
 - 要素技術調査でWebアプリのサンプルを作成していたときはclassesディレクトリにjavaファイルも置いていたのですが、コンパイル後のクラスファイルのみを置くのが正しいでしょうか?

![image](https://user-images.githubusercontent.com/54921025/72951832-2a94a580-3dd3-11ea-8978-5be413c40bf7.png)

wnakayama committed:
[シーケンス図]非同期通信部分の矢印,ライフライン表記を修正

yhanba commented:
参考情報ですが、
国税庁からは「一の適格請求書につき、税率ごとに１回の端数処理を行います。」
とあるので、「①単体の税込価格で切り捨てしてから足していく」という方法はダメですね。
https://www.nta.go.jp/taxes/shiraberu/zeimokubetsu/shohi/keigenzeiritsu/pdf/qa/01-09.pdf

yhanba commented:
機能仕様には、「商品情報は、最大で1ページあたり12商品分表示する」とあるので、
リクエストを受けた時に表示する商品を12個に絞る必要があると思うのですが、
それはどこで処理をするのでしょうか？
松村さんから商品の順番をソートする処理について指摘されているので、そこと合わせて考えれば良いと思います。
https://github.com/W-Nakayama-24/ojt_ecsite/pull/3#pullrequestreview-346300527

余裕があれば、13個目以降の商品を表示したい時の処理の流れが考えられていると良いと思います。

W-Nakayama-24 commented:
>国税庁からは「一の適格請求書につき、税率ごとに１回の端数処理を行います。」
とあるので、「①単体の税込価格で切り捨てしてから足していく」という方法はダメですね。

>機能仕様には、「商品情報は、最大で1ページあたり12商品分表示する」とあるので、
リクエストを受けた時に表示する商品を12個に絞る必要があると思うのですが、
それはどこで処理をするのでしょうか？
松村さんから商品の順番をソートする処理について指摘されているので、そこと合わせて考えれば良いと思います。

>余裕があれば、13個目以降の商品を表示したい時の処理の流れが考えられていると良いと思います。

レビューありがとうございます。
今回の課題で求められるQCDとの兼ね合いを考慮して、「適格請求書等保存方式にのっとった端数処理」と「13個目以降の商品を表示したい時の処理の流れ」はmust要件を一度全て満たした後に考えようと思います。

「リクエストを受けた時に表示する商品を12個に絞る処理」については、後ほどシーケンス図にも明記します。
